### DISCUSSION TITLE
A really simple webpack.config.js

### DISCUSSION THREAD

**Discussion by hrishikesh-k** [webpack/webpack#13286] [Q&A] [answered]
Hi all!

I'm new to Webpack and I'm going through the official documentation and various StackOverflow posts (which are kinda old), but can't find my way. I'd really appreciate some help or pointer on what exactly I need to do.

So, I have 2 files, a simple JS file that I need to export as UMD and a simple CSS file. I wish to export both of these files, minified in a folder. Seems pretty simple in words, but it's getting really hard for me to figure this out.

I was already generating the UMD file successfully like this:

```JS
module.exports = {
  mode: 'production',
  entry: './src/foo.js',
  output: {
    filename: 'foo.min.js',
    path: require('path').resolve(__dirname, 'dist'),
    library: {
      name: 'Foo',
      type: 'umd'
    }
  }
}
```

The problems began when I decided to add the CSS file too. Since CSS file is not natively supported without any loaders, I'm not sure how to work with them. Docs mention using this plugin, that plugin, but do I really need multiple? Just minification and copying a CSS file from source to destination seems a small task in theory.

How do I go about this? Any resources that I might have missed or that would help me figure this out would do too.

Thank you.

**Comment by alexander-akait** [comment, marked as answer]
Don't forget CSS can contain `url` and `import` and you need handle this, for just copy, you can use https://github.com/webpack-contrib/copy-webpack-plugin, anyway if you need good CSS support you need to use `css-loader` https://github.com/webpack-contrib/css-loader

**Reply by hrishikesh-k** [reply]
Hi, thanks a lot for the reply.

Yes, I'm aware of the existing loaders, I was looking for examples to export a JS file along with a CSS file. After more finding, I'm landed on this:

```JS
const path = require('path')

let config = {
  module: {}
}

let jsConfig = Object.assign({}, config, {
  mode: 'production',
  entry: './src/foo.js',
  output: {
    filename: 'foo.min.js',
    path: path.resolve(__dirname, 'dist'),
    library: {
      name: 'Foo',
      type: 'umd'
    }
  }
})

let cssConfig = Object.assign({}, config, {
  mode: 'production',
  entry: './src/bar.css',
  output: {
    filename: 'bar.min.css',
    path: path.resolve(__dirname, 'dist'),
  },
  module: {
    rules: [{
      test: /\.css$/i,
      loader: 'css-loader'
    }]
  }
})

module.exports = [jsConfig, cssConfig]
```

This seems to work to export my files just fine. I'm getting a JS and CSS file as needed. However, the CSS file seems to be a JS file which might be injecting styles later. This is what I got:

![image](https://user-images.githubusercontent.com/11403551/116787438-9fb85580-aac1-11eb-8a78-decfcee4d8c7.png)


I don't think I need all of this, all I need is a simple minified CSS file. I currently have a really plain CSS file with comments, I just wish to remove those comments and trim spaces, etc.

Have I done this at least 1% corrent?

**Reply by alexander-akait** [reply]
If you need just a copy CSS file you can use `copy-webpack-plugin`, don't forget to copy images/fonts/etc and setup https://github.com/webpack-contrib/mini-css-extract-plugin/

Here two cases:
-  you need apply styles to DOM, in this case you need css-loader + style-loader (CSS inside JS)/mini-css-extract-plugin (CSS as external file) https://github.com/webpack-contrib/mini-css-extract-plugin/#common-use-case
- you need just copy file, here you can use `copy-webpack-plugin`

**Reply by hrishikesh-k** [reply]
Yeah, I was not able to use `copy-webpack-plugin` because it would not minify the file, just copy it as it is. But, I just found it, it has a transform option which can change the content while copying. So, here's my new `webpack.config.js` for anyone who might need it:

```JS
const path = require('path')
const CleanCSS = require('clean-css')
const CopyPlugin = require('copy-webpack-plugin')

module.exports = {
  mode: 'production',
  entry: './src/foo.js',
  output: {
    filename: 'foo.min.js',
    path: path.resolve(__dirname, 'dist'),
    library: {
      name: 'Foo',
      type: 'umd'
    }
  },
  plugins: [
    new CopyPlugin({
      patterns: [{
        to: './bar.min.css',
        from: path.resolve(__dirname, 'src', 'opinionJS.css'),
        transform: content => (new CleanCSS({level: 2}).minify(content)).styles
      }]
    })
  ]
}
```

I really appreciate your time, sir. Have a good day!

**Reply by alexander-akait** [reply]
I recommend to use https://github.com/webpack-contrib/mini-css-extract-plugin/#common-use-case because it has more features and better for perf, `transform` will be work also, but it is more for custom simple transformations

**Reply by hrishikesh-k** [reply]
I believe (from what I saw in that plugin), it works for CSS imported into JS files? Does it work fine for separate CSS files? Or are you suggesting that I should import my CSS file in my JS file and then use that plugin?

Please don't mind my questions, I'm just learning and thus, any suggestions to make my workflow optimum are highly appreciated. Thank you for your time.
